src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -0,0 +1,616 @@
+        self._killed = []
+
+        # The output map, mapping between nodes and the reach of that node.
+        self._reaches = []
+        self._writes = []
+
+    @property
+    def uses(self):
+        """
+        :returns: the list of nodes using the value that the referenced symbol
+                  has before it is reassigned.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._uses
+
+    @property
+    def defsout(self):
+        """
+        :returns: the list of nodes that reach the end of the block without
+                  being killed, and therefore can have dependencies outside
+                  of this block.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._defsout
+
+    @property
+    def killed(self):
+        """
+        :returns: the list of nodes that represent the last use of an assigned
+                  variable. Calling next_access on any of these nodes will find
+                  a write that reassigns it's value.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._killed
+
+    @property
+    def is_basic_block(self):
+        """
+        :returns: whether the scope of this DefinitionUseChain is a basic
+                  block, i.e. whether it contains any control flow nodes.
+        :rtype: bool
+        """
+        # A basic block is a scope without any control flow nodes inside.
+        # In PSyclone, possible control flow nodes are IfBlock, Loop
+        # and WhileLoop.
+        for node in self._scope:
+            c_f_nodes = node.walk((IfBlock, Loop, WhileLoop))
+            if len(c_f_nodes) > 0:
+                return False
+        return True
+
+    def find_forward_accesses(self):
+        """
+        Find all the forward accesses for the reference defined in this
+        DefinitionUseChain.
+        Forward accesses are all of the References (or Calls) that read
+        or write to the symbol of the reference up to the point that a
+        write to the symbol is guaranteed to occur.
+        PSyclone only assumes a write to the symbol is guaranteed to occur
+        if it occurs outside of a control flow region.
+
+        :returns: the forward accesses of the reference given to this
+                  DefinitionUseChain
+        :rtype: list[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        # Setup the start and stop positions
+        save_start_position = self._start_point
+        save_stop_position = self._stop_point
+        # If there is no set start point, then we look for all
+        # accesses after the Reference.
+        if self._start_point is None:
+            self._start_point = self._reference_abs_pos
+        # If there is no set stop point, then any Reference after
+        # the start point can potentially be a forward access.
+        if self._stop_point is None:
+            self._stop_point = sys.maxsize
```

Comment:
I added this as a test and this works - its handled by:
```
            control_flow_nodes, basic_blocks = self._find_basic_blocks(
                self._scope
             )
```
because it only creates basic blocks for the computed scope, which is the Routine ancestor.